src/amp/lib/experiment.ts

```diff
@@ -0,0 +1,150 @@
+interface ExperimentModel<VariantData> {
+    sticky: boolean;
+    consentNotificationId?: string;
+    variants: {
+        [variantName: string]: VariantData;
+    };
+}
+
+type AmpModel = ExperimentModel<number>;
+
+type StyledExperimentModel = ExperimentModel<VariantConfig>;
+
+interface VariantConfig {
+    proportion: number;
+    style: string;
+}
+
+interface ExperimentCollection<ExperimentObject> {
+    [experimentName: string]: ExperimentObject;
+}
+
+export type AmpModelCollection = ExperimentCollection<AmpModel>;
+
+export type StyledModelCollection = ExperimentCollection<StyledExperimentModel>;
+
+type StylesCollection = ExperimentCollection<{ [variantName: string]: string }>;
+
+const extractFromVariants = (
+    variants: { [variantName: string]: VariantConfig },
+    fieldName: keyof VariantConfig,
+): { [variantName: string]: string | number } => {
+    return Object.entries(variants).reduce((acc, [variantName, config]) => {
+        acc = { ...acc, ...{ [variantName]: config[fieldName] } };
+        return acc;
+    }, {});
+};
+
+const isValidAmpModelVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: number } => {
+    return Object.values(variant).every(x => typeof x === 'number');
+};
+
+const isValidStyleVariant = (variant: {
+    [variantName: string]: string | number;
+}): variant is { [variantName: string]: string } => {
+    return Object.values(variant).every(x => typeof x === 'string');
+};
+
+export const extractExperimentModels = (
+    fullConfig: StyledModelCollection,
+): AmpModelCollection => {
+    return Object.entries(fullConfig).reduce(
+        (extractedModels, [experimentName, styledModel]) => {
+            const { variants, ...configWithoutVariants } = styledModel;
+            const proportions = extractFromVariants(variants, 'proportion');
+            if (isValidAmpModelVariant(proportions)) {
+                extractedModels[experimentName] = {
+                    ...configWithoutVariants,
+                    ...{
+                        variants: proportions,
+                    },
+                };
+            }
+            return extractedModels;
+        },
+        {} as AmpModelCollection,
+    );
+};
```

Comment:
- outer can be expression rather than return
- no need to create inner object with one key&value just to spread it
- rather than mutating the accumulator you could return a new object
```suggestion
): AmpModelCollection => Object.entries(fullConfig).reduce(
      (extractedModels, [experimentName, styledModel]) => {
          const { variants, ...configWithoutVariants } = styledModel;
          const proportions = extractFromVariants(variants, 'proportion');
          if (isValidAmpModelVariant(proportions)) {
              return {
                ...extractedModels,
                [experimentName]: {
                    ...configWithoutVariants,
                    variants: proportions,
                },
              };
          }
          return extractedModels;
      },
      {} as AmpModelCollection,
  );
```